torch/csrc/jit/interpreter.cpp

```diff
@@ -283,97 +592,189 @@ struct Instruction {
-        // unused inputs are given a false register -1 so that we never hold a
-        // reference to the tensor data, otherwise we would fail to clean them
-        // up since they do not have a last use at which to free them
-        int reg = input->uses().size() > 0 ? getOrAllocateRegister(input) : -1;
-        listInsert(stage.inputs, reg);
+
+  // jump when input is not 0
+  void createJumpNZ(int from_inst, int to_inst) {
+    auto & inst = instructions[from_inst];
+    JIT_ASSERT(inst.debug_name == kPlaceholder);
+    auto offset = relativeJump(from_inst, to_inst);
+    inst.callback = [offset](Stack & stack) {
+      auto t = tensor_as<int64_t>(pop(stack));
+      return (t != 0) ? offset : 0;
+    };
+    inst.debug_name = kJumpNZ;
+  }
+
+  void createJump(int from_inst, int to_inst) {
+    auto & inst = instructions[from_inst];
+    JIT_ASSERT(inst.debug_name == kPlaceholder);
+    auto offset = relativeJump(from_inst, to_inst);
+    inst.callback = [=](Stack & stack) {
+      return offset;
+    };
+    inst.debug_name = kJump;
+  }
+
+  void insertNodesFromBlock(Block* block) {
+    for(auto node : block->nodes()) {
+      switch(node->kind()) {
+        case kIf: {
+          // x = if c:
+          //   <then_block>
+          //   -> (vt)
+          // else:
+          //    <else_block>
+          //   -> (vf)
+
+          // turns into:
+          //   JumpNZ c, then
+          //   <else_block>
+          //   x = vf
+          //   Jump end
+          // then:
+          //   <then_block>
+          //   x = vt
+          // end:
+
+          // kPlaceholder instructions are replaced with branch instructions
+          // when the branch target locations are known
+          auto cond_branch = insertInstruction(kPlaceholder, node->inputs(), moveFlags(node), {});
+          auto then_block = node->blocks()[0];
+          auto else_block = node->blocks()[1];
+          insertNodesFromBlock(else_block);
+          insertAssign(else_block->outputs(), moveFlags(else_block), node->outputs());
+          auto jump = insertInstruction(kPlaceholder, {}, {}, {});
+          auto then_block_start = instructions.size();
+          insertNodesFromBlock(then_block);
+          insertAssign(then_block->outputs(), moveFlags(then_block), node->outputs());
+          createJump(jump, instructions.size());
+          createJumpNZ(cond_branch, then_block_start);
+        } break;
+        case kLoop: {
+          // o0 = while c i0
+          //        block 0: l0
+          //          <body>
+          //          -> (v0, v1)
+
+          // turns into:
+          // l0 = i0
+          // JumpZ c, end
+          // begin:
+          //   <body>
+          //   c, l0 = v0, v1
+          //   JumpNZ c, begin
+          // end:
+
+          auto body_block = node->blocks()[0];
+
+          // before assign op: stack: ... <cond> <loop-carried-depdencies>
+          insertAssign(node->inputs(), moveFlags(node), body_block->inputs());
```

Comment:
Don't `loop-carried-dependencies` include things like cond and trip count? Or is it because they are removed at this point?

Reply:
They are removed at this point.